src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -230,125 +285,150 @@ RouterExecutorRun(QueryDesc *queryDesc, ScanDirection direction, long count, Tas
  */
-static int32
-ExecuteDistributedModify(Task *task)
+static bool
+ExecuteTaskAndStoreResults(QueryDesc *queryDesc, Task *task,
+						   bool isModificationQuery,
+						   bool expectResults)
 {
-	int32 affectedTupleCount = -1;
+	TupleDesc tupleDescriptor = queryDesc->tupDesc;
+	EState *executorState = queryDesc->estate;
+	MaterialState *routerState = (MaterialState *) queryDesc->planstate;
+	bool resultsOK = false;
+	List *taskPlacementList = task->taskPlacementList;
 	ListCell *taskPlacementCell = NULL;
 	List *failedPlacementList = NIL;
 	ListCell *failedPlacementCell = NULL;
+	int64 affectedTupleCount = -1;
+	bool gotResults = false;
 
-	foreach(taskPlacementCell, task->taskPlacementList)
+	/*
+	 * Try to run the query to completion on one placement. If the query fails
+	 * attempt the query on the next placement.
+	 */
+	foreach(taskPlacementCell, taskPlacementList)
 	{
 		ShardPlacement *taskPlacement = (ShardPlacement *) lfirst(taskPlacementCell);
 		char *nodeName = taskPlacement->nodeName;
 		int32 nodePort = taskPlacement->nodePort;
+		bool queryOK = false;
+		int64 currentAffectedTupleCount = 0;
+		PGconn *connection = GetOrEstablishConnection(nodeName, nodePort);
 
-		PGconn *connection = NULL;
-		PGresult *result = NULL;
-		char *currentAffectedTupleString = NULL;
-		int32 currentAffectedTupleCount = -1;
-
-		Assert(taskPlacement->shardState == FILE_FINALIZED);
-
-		connection = GetOrEstablishConnection(nodeName, nodePort);
 		if (connection == NULL)
 		{
 			failedPlacementList = lappend(failedPlacementList, taskPlacement);
 			continue;
 		}
 
-		result = PQexec(connection, task->queryString);
-		if (PQresultStatus(result) != PGRES_COMMAND_OK)
+		queryOK = SendQueryInSingleRowMode(connection, task->queryString);
+		if (!queryOK)
 		{
-			ReportRemoteError(connection, result);
-			PQclear(result);
-
+			PurgeConnection(connection);
 			failedPlacementList = lappend(failedPlacementList, taskPlacement);
 			continue;
 		}
 
-		currentAffectedTupleString = PQcmdTuples(result);
-		currentAffectedTupleCount = pg_atoi(currentAffectedTupleString, sizeof(int32), 0);
-
-		if ((affectedTupleCount == -1) ||
-			(affectedTupleCount == currentAffectedTupleCount))
+		/*
+		 * If caller is interested, store query results the first time
+		 * through. The output of the query's execution on other shards is
+		 * discarded if we run there (because it's a modification query).
+		 */
+		if (!gotResults && expectResults)
 		{
-			affectedTupleCount = currentAffectedTupleCount;
+			queryOK = StoreQueryResult(routerState, connection, tupleDescriptor,
+									   &currentAffectedTupleCount);
 		}
 		else
 		{
-			ereport(WARNING, (errmsg("modified %d tuples, but expected to modify %d",
-									 currentAffectedTupleCount, affectedTupleCount),
-							  errdetail("modified placement on %s:%d",
-										nodeName, nodePort)));
+			queryOK = ConsumeQueryResult(connection, &currentAffectedTupleCount);
```

Comment:
Because we already used `queryOK` variable in the return value of `SendQueryInSingleRowMode()` and in an if statement. Using it for another purpose makes following the code a bit harder.